libraries/AP_Camera/AP_RunCam.h

```diff
@@ -0,0 +1,358 @@
+
+    // operation of 5 Key OSD cable simulation
+    enum class SimulationOperation
+    {
+        SIMULATION_NONE = 0x00,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_SET = 0x01,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_LEFT = 0x02,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_RIGHT = 0x03,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_UP = 0x04,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_DOWN = 0x05
+    };
+
+    // operation of RC5KEY_CONNECTION
+    enum class ConnectionOperation
+    {
+        RCDEVICE_PROTOCOL_5KEY_FUNCTION_OPEN = 0x01,
+        RCDEVICE_PROTOCOL_5KEY_FUNCTION_CLOSE = 0x02
+    };
+
+    // protocol versions, only version 1.0 is supported
+    enum class ProtocolVersion
+    {
+        RCSPLIT_VERSION = 0x00, // unsupported firmware version <= 1.1.0
+        VERSION_1_0 = 0x01,
+        UNKNOWN
+    };
+
+    // status of command
+    enum class RequestStatus
+    {
+        NONE,
+        PENDING,
+        SUCCESS,
+        INCORRECT_CRC,
+        TIMEOUT
+    };
+
+    enum class State {
+        INITIALIZING, // uart open
+        INITIALIZED,  // features received
+        READY,
+        VIDEO_RECORDING,
+        ENTERING_MENU,
+        IN_MENU,
+        EXITING_MENU
+    };
+
+    enum class Event {
+        NONE,
+        ENTER_MENU,
+        EXIT_MENU,
+        IN_MENU_ENTER,
+        IN_MENU_RIGHT, // only used by the 5-key process
+        IN_MENU_UP,
+        IN_MENU_DOWN,
+        IN_MENU_EXIT,
+        BUTTON_RELEASE,
+        STOP_RECORDING,
+        START_RECORDING
+    };
+
+    // initialize the RunCam driver
+    void init();
+    // camera button simulation
+    bool simulate_camera_button(const ControlOperation operation);
+    // start the video
+    void start_recording();
+    // stop the video
+    void stop_recording();
+    // update loop
+    void update();
+    // Check whether arming is allowed
+    bool pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) const;
+
+    static const uint32_t RUNCAM_MAX_PACKET_SIZE =       64;
+    static const uint8_t  RUNCAM_NUM_SUB_MENUS =          5;
+    static const uint8_t  RUNCAM_NUM_EXPECTED_RESPONSES = 4;
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    // supported features, usually probed from the device
+    AP_Int16 _features;
+    // number of initialization attempts
+    AP_Int8 _init_attempts;
+    // delay between initialization attempts
+    AP_Int32 _init_attempt_interval_ms;
+    // delay time to make sure the camera is fully booted
+    AP_Int32 _boot_delay_ms;
+    // delay time to make sure a button press has been activated
+    AP_Int32 _button_delay_ms;
+
+    // video on/off
+    bool _video_recording = true;
+    // detected protocol version
+    ProtocolVersion _protocol_version = ProtocolVersion::UNKNOWN;
+    // uart for the device
+    AP_HAL::UARTDriver *uart;
+    // camera state
+    State _state = State::INITIALIZING;
+    // count of ticks since last OSD cycle
+    uint32_t _tick_count;
```

Comment:
So I did it this way because decimation was easy and the rate of the OSD doesn't have to be super well controlled. If we start having to write a partial scheduler inside update() then I am against doing it this way - it's just duplicating somewhat tricky code code little obvious benefit.

Reply:
Ok fixed